### PR TITLE
fix: Decrease Address Generation Time

### PR DESCRIPTION
**Pull request by coreyphillips**
This PR:
- Implements `nodejs-mobile-react-native`.
- Wraps the following methods: `generateMnemonic`, `getPrivateKey`, `getScriptHash`, `getAddress`.
- The app will now generate 5 addresses and 5 changeAddresses on start.
  - Eventually we'll want to bump this to twenty to match the gap limit standard.
- Added logic to the `subscribeToAddresses` method to ensure we are watching behind and ahead of the current address index by the set gap limit or highest address count available if less than the current gap limit.

### CODE REVIEW

File: postinstall.js [hunk 1]
```diff
@@ -1,7 +1,11 @@
 var exec = require('child_process').exec;
 var os = require('os');
 
-const baseCommand = `rn-nodeify --install buffer,stream,assert,events,crypto,vm,process --hack`;
+const baseCommand = `
+cd nodejs-assets/nodejs-project && 
+npm i &&
```
Comment (limpbrains) [inline postinstall.js]: I think we need to stick to yarn here. And include yarn.lock

**Review by limpbrains — CHANGES_REQUESTED** [review]

File: postinstall.js (hunk 1 above)
Comment (coreyphillips) [reply postinstall.js]: Updated [here](https://github.com/synonymdev/bitkit/pull/368/commits/60d49fd7a33bdb71328f0814005097e5dd910b96).

**Review by limpbrains — APPROVED** [review]

**Comment by pwltr** [comment]
I'm seeing some untracked files on my machine after running `yarn`:

```
On branch nodejs-mobile-wrappers
Your branch is up to date with 'origin/nodejs-mobile-wrappers'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        nodejs-assets/build-native-modules-MacOS-helper-script-node.sh
        nodejs-assets/build-native-modules-MacOS-helper-script-npm.sh
        nodejs-assets/nodejs-project/sample-main.js
        nodejs-assets/nodejs-project/sample-package.json

nothing added to commit but untracked files present (use "git add" to track)
```

Can we add or ignore these?

**Comment by pwltr** [comment]
Builds are failing for me. Any ideas?

Android:
```
> Task :nodejs-mobile-react-native:GenerateNodeNativeAssetsListsarm64-v8a FAILED

FAILURE: Build failed with an exception.

* What went wrong:
A problem was found with the configuration of task ':nodejs-mobile-react-native:GenerateNodeNativeAssetsListsarm64-v8a' (type 'DefaultTask').
  - Property '$1' specifies directory '/Users/phil/Dev/synonymdev/bitkit/android/build/nodejs-native-assets/nodejs-native-assets-arm64-v8a' which doesn't exist.
    
    Reason: An input file was expected to be present but it doesn't exist.
```

iOS:
```
An error was encountered processing the command (domain=IXUserPresentableErrorDomain, code=1):
Unable To Install “BitKit”
Please try again later.
invalid symlink at /Users/phil/Library/Developer/CoreSimulator/Devices/DB6D0A67-87BE-42D8-BBB7-1608CD52ED48/data/Library/Caches/com.apple.mobile.installd.staging/temp.1pWpdr/extracted/bitkit.app/nodejs-project/node_modules/tiny-secp256k1/build/node_gyp_bins/python3
Underlying error (domain=MIInstallerErrorDomain, code=70):
        invalid symlink at /Users/phil/Library/Developer/CoreSimulator/Devices/DB6D0A67-87BE-42D8-BBB7-1608CD52ED48/data/Library/Caches/com.apple.mobile.installd.staging/temp.1pWpdr/extracted/bitkit.app/nodejs-project/node_modules/tiny-secp256k1/build/node_gyp_bins/python3
```

And `yarn start` gives me:
```
jest-haste-map: Haste module naming collision: bitkit-project
  The following files share their name; please adjust your hasteImpl:
    * <rootDir>/nodejs-assets/nodejs-project/package.json
    * <rootDir>/android/build/nodejs-native-assets-temp-build/nodejs-native-assets-arm64-v8a/nodejs-project/package.json

Failed to construct transformer:  DuplicateError: Duplicated files or mocks. Please check the console for more info
    at setModule (/Users/phil/Dev/synonymdev/bitkit/node_modules/jest-haste-map/build/index.js:576:17)
    at workerReply (/Users/phil/Dev/synonymdev/bitkit/node_modules/jest-haste-map/build/index.js:647:9)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Promise.all (index 140)
    at async /Users/phil/Dev/synonymdev/bitkit/node_modules/jest-haste-map/build/index.js:459:22
    at async DependencyGraph.load (/Users/phil/Dev/synonymdev/bitkit/node_modules/metro/src/node-haste/DependencyGraph.js:86:36)
    at async Bundler.ready (/Users/phil/Dev/synonymdev/bitkit/node_modules/metro/src/Bundler.js:61:5)
    at async IncrementalBundler.ready (/Users/phil/Dev/synonymdev/bitkit/node_modules/metro/src/IncrementalBundler.js:309:5)
    at async Server.ready (/Users/phil/Dev/synonymdev/bitkit/node_modules/metro/src/Server.js:1156:5) {
  mockPath1: 'nodejs-assets/nodejs-project/package.json',
  mockPath2: 'android/build/nodejs-native-assets-temp-build/nodejs-native-assets-arm64-v8a/nodejs-project/package.json'
```

**Comment by coreyphillips** [comment]
> I'm seeing some untracked files on my machine after running `yarn`:
> 
> ```
> On branch nodejs-mobile-wrappers
> Your branch is up to date with 'origin/nodejs-mobile-wrappers'.
> 
> Untracked files:
>   (use "git add <file>..." to include in what will be committed)
>         nodejs-assets/build-native-modules-MacOS-helper-script-node.sh
>         nodejs-assets/build-native-modules-MacOS-helper-script-npm.sh
>         nodejs-assets/nodejs-project/sample-main.js
>         nodejs-assets/nodejs-project/sample-package.json
> 
> nothing added to commit but untracked files present (use "git add" to track)
> ```
> 
> Can we add or ignore these?

Correct. I updated the gitignore [here](https://github.com/synonymdev/bitkit/pull/368/commits/4bc69c47ceb08b7bf4368a7404457e6de62005ea). Those are generated from the nodejs-mobile postinstall and can be safely ignored. The .sh files are specific to the machine and shouldn't be added, the sample files are example files used as a reference for new projects and not essential to the project.

**Comment by coreyphillips** [comment]
> Builds are failing for me. Any ideas?
> 
> Android:
> 
> ```
> > Task :nodejs-mobile-react-native:GenerateNodeNativeAssetsListsarm64-v8a FAILED
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> A problem was found with the configuration of task ':nodejs-mobile-react-native:GenerateNodeNativeAssetsListsarm64-v8a' (type 'DefaultTask').
>   - Property '$1' specifies directory '/Users/phil/Dev/synonymdev/bitkit/android/build/nodejs-native-assets/nodejs-native-assets-arm64-v8a' which doesn't exist.
>     
>     Reason: An input file was expected to be present but it doesn't exist.
> ```
> 
> iOS:
> 
> ```
> An error was encountered processing the command (domain=IXUserPresentableErrorDomain, code=1):
> Unable To Install “BitKit”
> Please try again later.
> invalid symlink at /Users/phil/Library/Developer/CoreSimulator/Devices/DB6D0A67-87BE-42D8-BBB7-1608CD52ED48/data/Library/Caches/com.apple.mobile.installd.staging/temp.1pWpdr/extracted/bitkit.app/nodejs-project/node_modules/tiny-secp256k1/build/node_gyp_bins/python3
> Underlying error (domain=MIInstallerErrorDomain, code=70):
>         invalid symlink at /Users/phil/Library/Developer/CoreSimulator/Devices/DB6D0A67-87BE-42D8-BBB7-1608CD52ED48/data/Library/Caches/com.apple.mobile.installd.staging/temp.1pWpdr/extracted/bitkit.app/nodejs-project/node_modules/tiny-secp256k1/build/node_gyp_bins/python3
> ```
> 
> And `yarn start` gives me:
> 
> ```
> jest-haste-map: Haste module naming collision: bitkit-project
>   The following files share their name; please adjust your hasteImpl:
>     * <rootDir>/nodejs-assets/nodejs-project/package.json
>     * <rootDir>/android/build/nodejs-native-assets-temp-build/nodejs-native-assets-arm64-v8a/nodejs-project/package.json
> 
> Failed to construct transformer:  DuplicateError: Duplicated files or mocks. Please check the console for more info
>     at setModule (/Users/phil/Dev/synonymdev/bitkit/node_modules/jest-haste-map/build/index.js:576:17)
>     at workerReply (/Users/phil/Dev/synonymdev/bitkit/node_modules/jest-haste-map/build/index.js:647:9)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async Promise.all (index 140)
>     at async /Users/phil/Dev/synonymdev/bitkit/node_modules/jest-haste-map/build/index.js:459:22
>     at async DependencyGraph.load (/Users/phil/Dev/synonymdev/bitkit/node_modules/metro/src/node-haste/DependencyGraph.js:86:36)
>     at async Bundler.ready (/Users/phil/Dev/synonymdev/bitkit/node_modules/metro/src/Bundler.js:61:5)
>     at async IncrementalBundler.ready (/Users/phil/Dev/synonymdev/bitkit/node_modules/metro/src/IncrementalBundler.js:309:5)
>     at async Server.ready (/Users/phil/Dev/synonymdev/bitkit/node_modules/metro/src/Server.js:1156:5) {
>   mockPath1: 'nodejs-assets/nodejs-project/package.json',
>   mockPath2: 'android/build/nodejs-native-assets-temp-build/nodejs-native-assets-arm64-v8a/nodejs-project/package.json'
> ```

You may need to clean your build with `yarn clean:all` or `git clone` and start from scratch. I also updated `package.json`'s clean script [here](https://github.com/synonymdev/bitkit/pull/368/commits/2458e121d00926560003ee88dc982f2fae2589dc) to include the `nodejs-mobile` `node_modules`.

**Comment by pwltr** [comment]
Getting the same build errors on a fresh `git clone`. What's your node version? I'm on `v18.8.0`

**Comment by coreyphillips** [comment]
Currently building with `v16.15.1`

**Comment by pwltr** [comment]
No luck either with Node 16. Are you on ARM? The Android error hints at an arch problem:

```
- Property '$1' specifies directory '/Users/phil/Dev/synonymdev/bitkit/android/build/nodejs-native-assets/nodejs-native-assets-arm64-v8a' which doesn't exist.
```

**Comment by coreyphillips** [comment]
I was able to reproduce this error on an arm and intel machine by git cloning and installing with node v18 set. I then switched to node v16 and ran `yarn clean && yarn ios`. That yielded the build error for me. So it doesn't seem to like switching node versions once you've `yarn install`ed.

For a successful build I suggest erasing all contents and settings in your simulator (Device->Erase All Content And Settings), set node to either v18 or v16, move into a new, empty directory and run `git clone git@github.com:synonymdev/bitkit.git && cd bitkit && git checkout nodejs-mobile-wrappers && yarn install`. Once installed, running `yarn ios` should yield a successful build.

It's possible there's some caching issue at play. I was able to create a successful build using both versions of node on an intel and arm-based machine.

**Comment by limpbrains** [comment]
iOS: 
- compiles, but shows error when I runs it:
```
ERROR  TypeError: Cannot read property 'startNodeProject' of null, js engine: hermes
 ERROR  Invariant Violation: Module AppRegistry is not a registered callable module (calling runApplication). A frequent cause of the error is that the application entry file path is incorrect.
      This can also happen when the JS bundle is corrupt or there is an early initialization error when loading React Native., js engine: hermes
 ERROR  Invariant Violation: Module AppRegistry is not a registered callable module (calling runApplication). A frequent cause of the error is that the application entry file path is incorrect.
      This can also happen when the JS bundle is corrupt or there is an early initialization error when loading React Native., js engine: hermes
```

Andoid:
I have the same error as @pwltr:
```
- Property '$1' specifies directory '/Users/phil/Dev/synonymdev/bitkit/android/build/nodejs-native-assets/nodejs-native-assets-arm64-v8a' which doesn't exist.
```

**Comment by limpbrains** [comment]
Also Metro shows me this error:
```
jest-haste-map: Haste module naming collision: bitkit-project
  The following files share their name; please adjust your hasteImpl:
    * <rootDir>/android/build/nodejs-native-assets-temp-build/nodejs-native-assets-armeabi-v7a/nodejs-project/package.json
    * <rootDir>/android/build/nodejs-native-assets/nodejs-native-assets-arm64-v8a/nodejs-project/package.json
```

name in nodejs-assets/nodejs-project/package.json needs to be changed

**Comment by coreyphillips** [comment]
> Also Metro shows me this error:
> 
> ```
> jest-haste-map: Haste module naming collision: bitkit-project
>   The following files share their name; please adjust your hasteImpl:
>     * <rootDir>/android/build/nodejs-native-assets-temp-build/nodejs-native-assets-armeabi-v7a/nodejs-project/package.json
>     * <rootDir>/android/build/nodejs-native-assets/nodejs-native-assets-arm64-v8a/nodejs-project/package.json
> ```
> 
> name in nodejs-assets/nodejs-project/package.json needs to be changed

This should be fixed [here](https://github.com/synonymdev/bitkit/pull/368/commits/976c40c8743efcc6c4e3adf62b1f32eda8973782) and may also resolve some of the other errors being mentioned.

Please use [these](https://github.com/synonymdev/bitkit/pull/368#issuecomment-1246725258) build instructions when testing a new build to rule out other issues.

**Comment by pwltr** [comment]
I got the iOS build working without issues, may have been `yarn cache clean` that ultimately helped. But Android I can't get to run at all.

**Comment by coreyphillips** [comment]
Can confirm. Something is blocking the Android build. Will do some testing and tweaking of the configs to hone in on it.

**Comment by coreyphillips** [comment]
@pwltr @limpbrains, Please confirm that you are able to run this on Android as of [this commit](https://github.com/synonymdev/bitkit/pull/368/commits/7c3124c0b98bf3f71b1685c24686aa32f013f97a).

**Comment by limpbrains** [comment]
Android works fine for me, can't test iOS right now

**Review by limpbrains — APPROVED** [review]

**Comment by pwltr** [comment]
I cannot get it to run on Android, still getting the same error as before. Tried both node `v16.17.0` and `v18.8.0`, each time starting from a fresh`git clone` and cleaning every cache I'm aware of. This is on an Intel x86.

**Comment by coreyphillips** [comment]
How are you attempting to run it? Currently, I will spin up the emulator in Android studio and run `yarn android` from the terminal. If that gives any trouble, clicking the "Run App" (Play) button in Android Studio should get it done.

@Jasonvdb does this build give you trouble on Android as well?

**Comment by pwltr** [comment]
To be exact, for me it fails at the Android build step. So I don't reach the point where I can try to run the build, be it in Android Studio or command-line. Would be good to get more eyes on this before I try to get this working again.

**Comment by coreyphillips** [comment]
Closing this for now so-as to not break or disrupt testing and debugging on Android. I intend to re-open this in ~2-3-weeks.

**Pull request closed without merging**